src/server/set_family.cc

```diff
@@ -284,19 +284,32 @@ void InterStrSet(const DbContext& db_context, const vector<SetType>& vec, String
   }
 }
 
-StringVec PopStrSet(const DbContext& db_context, unsigned count, const SetType& st) {
+StringVec RandMemberStrSet(const DbContext& db_context, const SetType& st,
+                           const std::unique_ptr<PicksGenerator>& generator,
+                           std::size_t picks_count) {
+  std::unordered_map<RandomPick, std::uint32_t> times_index_is_picked;
+  for (std::size_t i = 0; i < picks_count; i++) {
+    times_index_is_picked[generator->Generate()]++;
+  }
+
   StringVec result;
+  result.reserve(picks_count);
 
-  if (true) {
-    StringSet* ss = (StringSet*)st.first;
-    ss->set_time(MemberTimeSeconds(db_context.time_now_ms));
+  StringSet* ss = static_cast<StringSet*>(st.first);
+  ss->set_time(MemberTimeSeconds(db_context.time_now_ms));
 
-    // TODO: this loop is inefficient because Pop searches again and again an occupied bucket.
-    for (unsigned i = 0; i < count && !ss->Empty(); ++i) {
-      result.push_back(ss->Pop().value());
+  std::uint32_t ss_entry_index = 0;
+  for (const sds ptr : *ss) {
+    std::uint32_t t = times_index_is_picked[ss_entry_index++];
+    while (t--) {
+      result.emplace_back(ptr, sdslen(ptr));
     }
```

Comment:
you use an unordered_map, but during iteration you'll end up creating ever cell up to N because of `[ss_entry_index]`. We can use find() or use a vector because we currently iterate over the whole set either way